src/measure_theory/measure/measure_space.lean

```diff
@@ -2020,6 +2020,49 @@ lemma self_mem_ae_restrict {s} (hs : measurable_set s) : s ∈ (μ.restrict s).a
 by simp only [ae_restrict_eq hs, exists_prop, mem_principal, mem_inf_iff];
   exact ⟨_, univ_mem, s, subset.rfl, (univ_inter s).symm⟩
 
+/-- If two measurable sets are ae_eq then any proposition that is almost everywhere true on one
+is almost everywhere true on the other -/
+lemma restrict_eq_of_ae_eq_of_restrict_eq {s t} (hs : measurable_set s) (ht : measurable_set t)
+  (hst : s =ᵐ[μ] t) {p : α → Prop} :
+  (∀ᵐ (x : α) ∂μ.restrict s, p x) → (∀ᵐ (x : α) ∂μ.restrict t, p x) :=
+begin
+  rw [ae_restrict_iff' hs, ae_restrict_iff' ht, ae_iff, ae_iff],
+  simp only [not_forall, exists_prop],
+  intros h,
+  have aa : {a : α | a ∈ s ∧ ¬p a} ∪ {a : α | a ∈ t ∧ ¬p a} =
+    {a : α | a ∈ s ∧ ¬p a} ∪ {a : α | a ∈ (t \ s) ∧ ¬p a},
+  { ext,
+    simp only [mem_union_eq, mem_set_of_eq, mem_diff],
+    conv
+    { congr, congr, rw and_comm, skip, rw and_comm, skip, congr, rw and_comm, skip, rw and_comm, },
+    rw [←and_or_distrib_left, ←and_or_distrib_left, and.congr_right_iff],
+    intros hf,
+    rw [←mem_union, ←mem_diff, ←mem_union],
+    conv { to_rhs, rw union_comm, rw diff_union_self, rw union_comm, }, },
+  have bb : {a : α | a ∈ t ∧ ¬p a} ⊆ {a : α | a ∈ s ∧ ¬p a} ∪ {a : α | a ∈ t ∧ ¬p a}, simp,
+  have cc : μ ({a : α | a ∈ s ∧ ¬p a} ∪ {a : α | a ∈ t ∧ ¬p a}) = 0,
+  { rw aa,
+    have : μ ({a : α | a ∈ s ∧ ¬p a} ∪ {a : α | a ∈ (t \ s) ∧ ¬p a}) ≤
+      μ {a : α | a ∈ s ∧ ¬p a} + μ {a : α | a ∈ (t \ s) ∧ ¬p a},
+    { exact measure_union_le _ _, },
+    have dd : {a : α | a ∈ (t \ s) ∧ ¬p a} ⊆ t \ s,
+    { rw subset_def,
+      simp only [mem_diff, mem_set_of_eq, and_imp],
+      exact λ _ ht hs _, ⟨ht, hs⟩, },
+    rw ae_eq_set at hst,
+    rw [h, measure_mono_null dd hst.right, zero_add, nonpos_iff_eq_zero] at this,
+    exact this, },
+  exact measure_mono_null bb cc,
```

Comment:
```suggestion
  rw [ae_restrict_iff' hs, ae_restrict_iff' ht, ae_iff, ae_iff], intro h,
  let Q := {a : α | ¬p(a)}, 
  suffices : μ(t ∩ Q) = 0, { simpa using this },
  have : μ(s ∩ Q) = 0, { simpa using h },
  have aa : (s ∩ Q) ∪ (t ∩ Q) = (s ∩ Q) ∪ ((t \ s) ∩ Q),
  { rw [←inter_distrib_right, ←inter_distrib_right, union_diff_self] },
  have bb : (t ∩ Q) ⊆ (s ∩ Q) ∪ (t ∩ Q) := by simp,
  refine measure_mono_null bb _,
  rw aa, exact measure_union_null this (measure_mono_null (by simp) (ae_eq_set.mp hst).2),
```
(maybe this can be golfed even further)